### ISSUE TITLE
Training Maxing out very large memory. Slow and/or never starting...

### ISSUE THREAD

**Issue by brianfarrar** [matterport/Mask_RCNN#278]
Like several others I am trying to use the matterport/Mask_RCNN in the Data Science Bowl 2018.  I am running training in Google Cloud ML.  I adapted the "shapes" demo to run in Google Cloud and it runs fast and well.

I took the running Shapes adaptation and replaced the Config and Dataset with new definitions for the DSB 2018 data sets.   If I try to run the full datasets.  The model hangs somewhere after the call to model.train(...) and never comes back (at least I haven't waited longer than 2 hours for it start the first epoch).  When I scale the data down to 5 images in the train set and 2 in the validation set.  It runs the first epoch, and hangs in the 3rd step of the second epoch.  (Note the ~30 min period I gave it to move to step 4 before killing the job).

![image](https://user-images.githubusercontent.com/26047294/36646057-e9368556-1a37-11e8-8430-ebee6b9f42cc.png)

I note from the performance data that my master is running at 100% memory usage on scale tier complex_l_gpu which is the largest available machine in Google Cloud ML.  I have read @waleedka note on issue #116 which indicates that the sparse to dense tensor conversion is not an issue, but I am wondering if it may be...

I am using mini masks.  Here is the data from my config.  

![image](https://user-images.githubusercontent.com/26047294/36646156-9533fee6-1a39-11e8-89e1-4ed03a066a72.png)

I have tried it with 8 GPUs, I have tried it with 1.  I have tried scaling down the images.  I am currently setting up a standalone instance to eliminate the build in Google Cloud ML as the issue. (Will report back with results). Update, I have tested it on a stand alone instance as well.  With 208GB of memory, it is still maxing out!  After 90 min still waiting for the first epoch to start.  Something definitely wrong.  Could be my code, but I don't know how to diagnose.

I am wondering if anyone has a suggestion for diagnosing...

**Comment by brianfarrar** [comment]
Before I get into this post, I just want to say thanks to the authors for a very understandable implementation and I am looking forward to working with it.  Your work is appreciated!

On to this issue: so I have gotten this working.  I cannot definitively point to the cause, but here is some evidence for the record: 

1) in reviewing the other issues here raised, I noted issue #102 which is similar to the error message I was getting when I stopped my hung sessions (as noted above).  So I loaded up the inspect_data.ipynb with my config and class.  It ran end to end with no problem.  Including using the next generator to try reading every record.  (Very handy analysis by the way.  I recommend it for anyone having trouble.)  So I eliminated my code and data as the problem.

2) Next I set up a step debugging harness in the actual environment to step through to where I was hanging, and after seeing it hang in the same place 3 times (line 74 in python3.5/multiprocessing/popen_fork.py).  See below:

![image](https://user-images.githubusercontent.com/26047294/36767422-04a8728a-1c00-11e8-9b4d-1ab19696457a.png)

3) I recalled having stepped through the section where the number of workers is calculated.  There was a carve out for NT, where workers is set to 0, so I switched it to 0 for the case when the operating system is not NT as shown here:

![image](https://user-images.githubusercontent.com/26047294/36757773-0729f542-1bd8-11e8-823b-3b6f684138df.png)

I made no other changes and completed 10 full epochs in ~ 40 min.  In line with my expectations given the machine I was running on.  Here is the config I am running:

python 3.5.2
keras 2.1.4
tensorflow 1.4.1
ubuntu 16.04

I can't rule out that there is some interaction of version differences, or other bug, but seems like the multiprocessor issue noted in #13 may be either misdiagnosed, or otherwise applicable to other operating systems.  

My next step will be to make sure all 8 gpus I have are being used and if not, try taking the single model and then instantiating with keras.utils.multi_gpu_model(model, gpus=None).
 
I am going to close this issue for now.

**Comment by rida1896** [comment]
@brianfarrar 
Please can you tell me how you used the GPUs in google cloud. It gives me following error : "Permission denied (HTTP 403): Quota 'NVIDIA_K80_GPUS' exceeded. Limit: 0.0 in region asia-east1." 
Also, are you using the free version or paying for it? 
How did you fix this issue?

**Comment by brianfarrar** [comment]
That message is telling you that your account is not authorized to use GPUs.  You have to request GPU quota be added to your account.  To do this go to your google cloud console.  Select IAM|Quota select the check box labeled: Compute Engine API - NVIDIA K80 GPUs and the click "Edit Quotas".

I was using a paid account.  However, Google offers several hundred dollars of free credits to new users.  

Good luck!

Brian.
